### ISSUE TITLE
plugins should receive readOnly

### ISSUE THREAD

**Issue by macrozone** [globocom/megadraft#86]
Plugins should receive readOnly of the parent editor.

With this, you could easily use the editor to render the result and change the appearance of the plugin's component.

E.g. an image could show a form on readOnly=false and render just as an `<img />` tag on readOnly

**Comment by macrozone** [comment]
ok, i tried that out, but there is problem:

MediaWrapper does set readonly to true when it gains focus, so this will destroy this idea.

So i'll take a different approach...

**Comment by ricobl** [comment]
You're right @macrozone, this is somehow related to #82.

We should have a way to differentiate the original `readOnly` from the `readOnly` managed by `MediaWrapper`.

Please let us know if you need help.

**Comment by macrozone** [comment]
what i currently do is the following:

```
  <MegadraftEditor
        key={isEditing} // force rerender
        actions={options.actions}
        plugins={options.plugins.map(plugin => plugin({ isEditing }))}
        readOnly={!canEdit || !isEditing}
        editorState={editorState}
        blockRenderMap={options.blockRenderMap}
        onChange={(state) => {
          setEditorState(state);
        }}
      />

```

so my plugin-definitions are all functions that return the plugin-definition based on whether the editor is in "editmode" (isEditing = true) or not. 

I then built a small plugin system on top of megadraft's system (it uses a schema-based dialog built with https://github.com/vazco/uniforms/)

```
// usage:
export default createPlugin({
  type: 'image',
  Component,
  schema,
});


// implementation

import React from 'react';
import PluginEditableComponent from '../containers/plugin_editable_component';
import PluginButton from '../containers/plugin_button';

export default ({ type, schema = null, Component }) => ({ isEditing }) => {
  const blockComponent = ({ data, container }) => {
    const component = <Component data={data} container={container} />;
     // show the render-component when not in editmode or if the plugin has no schema
    if (!isEditing) {
      return component;
    }
    return (
      <PluginEditableComponent
        schema={schema}
        data={data}
        container={container}
      >
        {component}
      </PluginEditableComponent>
    );
  };
  const buttonComponent = props => <PluginButton type={type} {...props} />;

  return {
     // A unique plugin name used to identify the plugin and its blocks
    type,
     // React component to be rendered in the block sidebar
    buttonComponent,
     // React component for rendering the content block
    blockComponent,
  };
};


```

**Comment by ricobl** [comment]
Hey @macrozone! A very clean approach, great job!

It seems that for most people using the editor is the preferred way to render the "final" content. So maybe having a preview mode like yours could come in handy and could be better than using `readOnly` for the preview mode. Read-only and preview have different semantics.

On the other hand, we don't use the actual editor for this, instead we serialize the editor state and render the content on a backend application. This way we get the benefits of SEO on our pages and we have a lightweight frontend that doesn't include React, DraftJS or Megadraft. How do you handle this? Do you face the same issues?

**Comment by macrozone** [comment]
@ricobl : 

Having the same logic rendering the final component as the editor has benefits (first: simplicity). You could also use react to render the component on the server side. The dom-tree is a bit bloated at the moment, but that could be handled. "Performance last".

Implementing a different react-component to render the editor-state should not be too hard, but why implementing the wheel twice? I would love to have draftjs or megadraft providing a way to render the actual component, because everything is already there.

**Comment by ricobl** [comment]
Hey @macrozone,

Sorry for the delay, I had to put a little more thought into it before answering.

You have a good point, but I'm not sure this a responsibility the editor should have.

If we embrace the preview mode, plugin authors will be forced to maintain the
HTML version of the plugin and from my experience the default output will
never be enough, people will always ask for an extra class, attribute or div
somewhere.

We will write about our vision for the project in more detail but the short
version is that the editor is not intended to be a WYSIWYG, mostly due to the
variety of channels the content can be distributed: smartphones, desktops,
native apps, Google's AMP, Facebooks's Instant Articles, etc.

With this in mind, I believe we could have an alternative library to handle the rendering,
giving more control over the markup and removing the clutter required by the
editor.

Thank you very much for the rich discussion.

Please let me know if this works for you!

**Comment by ricobl** [comment]
Thanks for the fix @cameackeSEL! Great job!

It's released under the `0.4.18` version.

We're lacking docs about `blockComponent` props, I've created an issue for this: #137.
